benchmark/test/test_httpapi.py

```diff
@@ -0,0 +1,260 @@
+
+        self.backend = InMemoryBackend()
+        api = BenchmarkAPI_V1(self.backend)
+        site = server.Site(api.app.resource())
+
+        def make_client(listening_port):
+            addr = listening_port.getHost()
+            self.agent = client.ProxyAgent(
+                endpoints.TCP4ClientEndpoint(
+                    reactor,
+                    addr.host,
+                    addr.port,
+                ),
+                reactor,
+            )
+
+        listening = Deferred()
+        listening.addCallback(make_client)
+        endpoint = TestEndpoint(reactor, listening)
+        self.service = StreamServerEndpointService(endpoint, site)
+        self.service.startService()
+        return listening
+
+    def tearDown(self):
+        super(BenchmarkAPITests, self).tearDown()
+        return self.service.stopService()
+
+    def submit(self, result):
+        """
+        Submit a result.
+        """
+        json = dumps(result)
+        body = StringProducer(json)
+        req = self.agent.request("POST", "/benchmark-results",
+                                 bodyProducer=body)
+        return req
+
+    def check_response_code(self, response, expected_code):
+        self.assertEqual(
+            response.code, expected_code, "Incorrect response code")
+        return response
+
+    def parse_submit_response_body(self, body):
+        data = loads(body)
+        self.assertIn('version', data)
+        self.assertEqual(data['version'], 1)
+        self.assertIn('id', data)
+        return data['id']
+
+    def test_submit_success(self):
+        """
+        Valid JSON can be successfully submitted.
+        """
+        req = self.submit(self.RESULT)
+        req.addCallback(self.check_response_code, http.CREATED)
+        return req
+
+    def test_submit_response_format(self):
+        """
+        Returned content is the expected JSON.
+        """
+        req = self.submit(self.RESULT)
+        req.addCallback(client.readBody)
+        req.addCallback(self.parse_submit_response_body)
+        return req
+
+    def test_submit_response_location_header(self):
+        """
+        Returned Location header has the expected value.
+        """
+        # This is not a real array, but a well-known trick
+        # to set an outer variable from an inner function.
+        response = [None]
+        req = self.submit(self.RESULT)
+
+        def save_response(_response):
+            response[0] = _response
+            return _response
+
+        req.addCallback(save_response)
+        req.addCallback(client.readBody)
+        req.addCallback(self.parse_submit_response_body)
+
+        def check_location(id):
+            expected_location = urljoin(
+                response[0].request.absoluteURI + '/',
+                id
+            )
+            self.assertEqual(
+                expected_location,
+                response[0].headers.getRawHeaders(b'Location')[0]
+            )
```

Comment:
I have almost no experience with testtools, so I am not sure what kind of a matcher would make most sense here.
E.g.:
* `assertThat(response, hasExpectedHeader(name, value))`
* `assertThat(response, hasExpectedLocation(location))`
* `assertThat(response, hasConsistentLocationAndID())`

Or something else?